### ISSUE TITLE
Predicting new instances

### ISSUE THREAD

**Issue by annm802** [MaartenGr/BERTopic#1480]
Hello ,

I'm trying to use .transform() to predict topics for my test set (the test set has a length of 1863). I'm however getting predictions of the data I used to train my model, instead of the new dataset. The train set has a length of 91596, which is equal to the length of the predicted topics.

It was working well last week with a different dataset though. 
What could be the issue?

**Comment by MaartenGr** [comment]
`.transform` will predict topics for any input you give it. This should mean that if you give it your test set, the output will also be the predictions of that test set. 

> The train set has a length of 91596, which is equal to the length of the predicted topics.

Are you sure you did not give the train set to the `.transform` function?

Could you share your full code? Which version of BERTopic are you using?

**Comment by annm802** [comment]
Hi @MaartenGr, thanks for the quick response.
I'm using bertopic==0.15.0
Here is my code.
```
import locale
locale.getpreferredencoding = lambda: "UTF-8"
!pip install --extra-index-url https://pypi.nvidia.com cuml-cu11
!pip install cudf-cu11 dask-cudf-cu11 --extra-index-url=https://pypi.nvidia.com
!pip install cuml-cu11 --extra-index-url=https://pypi.nvidia.com
!pip install cugraph-cu11 --extra-index-url=https://pypi.nvidia.com
!pip install cupy-cuda11x -f https://pip.cupy.dev/aarch64

from google.colab import drive
drive.mount('/content/drive')

import locale
locale.getpreferredencoding = lambda: "UTF-8"
!pip install -r '/content/drive/MyDrive/requirements.txt'

from cuml.cluster import HDBSCAN
from sentence_transformers import SentenceTransformer

from bertopic import BERTopic
from bertopic.cluster import BaseCluster
from bertopic.representation import KeyBERTInspired


import collections
from cuml.manifold import UMAP
from tqdm import tqdm
from sklearn.feature_extraction.text import CountVectorizer

import pandas as pd
import re
import numpy as np
import os

#preliminary imports
import sys
import operator
sys.path.append('/content/drive/MyDrive/')
import dimensionality
from dimensionality import *
from crunch_preprocessing import crunch_cleaning, no_white_spaces


#The data set
df = pd.read_csv('/content/drive/MyDrive/posts.csv')

#Getting the train and test sets
train_set = posts.sample(frac = 0.98,random_state= 42)
val_set = posts.drop(train_set.index)

train_set = train_set.reset_index(drop= True)
val_set = val_set.reset_index(drop= True)


val_set.to_csv('/content/drive/MyDrive/techcrunch_val_set.csv')


#BERTOPIC
docs = train_set['content']
#crunch_boom = post_booms['content']
# # # Create embeddings
model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')
crunch_embeddings = model.encode(docs, show_progress_bar=True)


with open('/content/drive/MyDrive/crunch_embeddings.npy', 'wb') as f:
       np.save(f, crunch_embeddings)

crunch_embeddings = np.load('/content/drive/MyDrive/crunch_embeddings.npy')

# Extract vocab to be used in BERTopic
crunch_vocab = collections.Counter()
tokenizer = CountVectorizer().build_tokenizer()
for doc in tqdm(docs):
  crunch_vocab.update(tokenizer(doc))
crunch_vocab = [word for word, frequency in crunch_vocab.items() if frequency >= 15]; len(crunch_vocab)

# Using UMAP to reduce dimensionality of embeddings
crunch_umap_model = UMAP(n_components=3, n_neighbors=15, learning_rate= 1e-09, random_state=42, metric="cosine", verbose=True)
reduced_crunch_embeddings = crunch_umap_model.fit_transform(crunch_embeddings)

# Finding clusters of lower-dimension embeddings to find similar posts
crunch_hdbscan_model = HDBSCAN(min_samples=45, gen_min_span_tree=True, prediction_data=True, min_cluster_size=20, verbose=True)
crunch_clusters = crunch_hdbscan_model.fit(reduced_crunch_embeddings).labels_

# Prepare sub-models
embedding_model = SentenceTransformer('all-MiniLM-L6-v2')
crunch_umap_model =dimensionality.Dimensionality(reduced_crunch_embeddings)
cruncht_hdbscan_model = BaseCluster()
vectorizer_model = CountVectorizer(vocabulary=crunch_vocab, stop_words="english")
representation_model = KeyBERTInspired()

# Fit BERTopic without actually performing any clustering
crunch_topic_model= BERTopic(
        embedding_model=embedding_model,
        umap_model=crunch_umap_model,
        hdbscan_model=crunch_hdbscan_model,
        vectorizer_model=vectorizer_model,
        representation_model=representation_model,
        verbose=True
).fit(docs, embeddings=crunch_embeddings, y=crunch_clusters)

crunch_topics_df = crunch_topic_model.get_topic_info()



new_dict= {'other':[-1,16,23,26,35,42,59,74,83,84,85,87,97,115,120,124,141,151,15,39,61,78,90,134,161,163],'vehicles & transportation':[0,110],
           'ventures, acquisitions, stocks, earnings ':[1,37,40,53,57,58,62,66,70,82,88,118,135,144,148,156,157], 'voting':[145],
           'social networks':[2,11,30,54,55,93,94,102,113,125,131,137], 'quantum computing':[106],
           'OS & browsers':[31,52,119,123], 'photos & videos':[92], 'digital payments':[114],
           'fintech':[3,73], 'music': [4,63], 'space exploration': [6], 'games':[7,68,143,152],
           'regulations,lawsuits & cybersecurity':[36,104,105,112,28,32,46,71,99,109,128,129,136,140,149,153],
           'streaming,movies & shows':[8,12,98,126,139,142], 'servers':[9,49,162], 'VR':[10,154], 'deliveries': [13,103], 'blockchain':[14,75],
           'energy':[146],
           'tech devices & accessories':[17,19,21,22,27,33,34,38,41,47,50,60,67,69,76,81,95,107,108,121,122,127,133,138,147,150,159,164],
           'renting & travel':[77,44,64,91],
           'telehealth':[18], 'AI & robotics':[20,29,51,89,111], 'retail & shopping':[24,100,116,132], 'education':[25],
           'storage':[43,80], 'employment & diversity':[45,56,96,160],'dating apps': [48], 'language & speech':[65,101,131]}

topics_to_merge = [a_list for a_list in new_dict.values()]


crunch_topic_model.merge_topics(docs, topics_to_merge= topics_to_merge)

crunch_topic_model.save('/content/drive/MyDrive/crunch_32topics_model', serialization="pickle")
topic_model = BERTopic.load('/content/drive/MyDrive/crunch_32topics_model')

new_topics_to_merge= [[8,31],[-1,10,27],[24,28]]
topic_model.merge_topics(docs, topics_to_merge= new_topics_to_merge)

topic_model.set_topic_labels({-1:'other', 0:'vehicles & transportation', 1:'tech devices & accessories',
                              2:'ventures,acquisitions & stocks ',3:'social networks',4:'streaming,movies & shows', 5:'fintech',
                              6:'regulations,lawsuits & cybersecurity',7:'music',8:'games',9:'space exploration',
                              10:'AI & robotics', 11:'servers',12:'VR',13:'blockchain',14:'deliveries',
                               15:'OS & browsers', 16:'retail & shopping', 17:'telehealth', 18:'renting & travel',
                               19:'employment & diversity',20:'education', 21:'storage',22:'language & speech',
                              23:'dating apps',24:'photos & videos',25:'quantum computing',26:'voting',27:'energy'})

tdf=topic_model.get_topic_info()


topic_model.save('/content/drive/MyDrive/crunch_27topics_model', serialization="pickle")



val_set= pd.read_csv('/content/drive/MyDrive/techcrunch_val_set.csv')
val_set= val_set[['year','title','content']]


#Making predictions
test_docs= val_set['content']
predicted_topics, probs = topic_model.transform(test_docs)`
```

**Comment by MaartenGr** [comment]
From a quick glance, I cannot see any problems. Also, when closely inspecting the `.transform` code internally, there is no mention of returning or using the topics of documents the model was trained on. My guess is that there is something going on with passing your dataset. Before you run the following:

```python
predicted_topics, probs = topic_model.transform(test_docs)
```

Could you check the length of `test_docs` and the length of `predicted_topics` after running the code?

**Comment by annm802** [comment]
Here is a screenshot of the lengths of `test_docs` and `predicted_topics`
![image](https://github.com/MaartenGr/BERTopic/assets/57720741/89885575-ec34-479a-9d27-37ef507b35a9)

I'm also getting the same result when predicting just one new instance, `x`
![image](https://github.com/MaartenGr/BERTopic/assets/57720741/65f31ade-b434-45ef-b09d-e4c52839d467)

**Comment by MaartenGr** [comment]
Hmmm, I am not sure what is happening here... Could you check the shape of `topic_model.topic_embeddings_`? Maybe they were saved incorrectly. Also, did you make sure to keep the environments between saving and loading exactly the same? 

Other than that, it might be worthwhile to use `safetensors` or `pytorch` for serialization instead. Generally, that is a cleaner approach.

**Comment by annm802** [comment]
> Hmmm, I am not sure what is happening here... Could you check the shape of `topic_model.topic_embeddings_`? Maybe they were saved incorrectly. Also, did you make sure to keep the environments between saving and loading exactly the same?
> 
> Other than that, it might be worthwhile to use `safetensors` or `pytorch` for serialization instead. Generally, that is a cleaner approach.

So I reduced the topics to 25, and I'm now getting a shape of (26 , 384). The environments are the same. I also used the same code for a different dataset, and the predictions worked fine.

I'm also using this:
````
class Dimensionality:
  """ Use this for pre-calculated reduced embeddings """
  def __init__(self, reduced_embeddings):
    self.reduced_embeddings = reduced_embeddings

  def fit(self, X):
    return self

  def transform(self, X):
    return self.reduced_embeddings`
```

**Comment by MaartenGr** [comment]
Looking through your code, I noticed you passed `crunch_hdbscan_model` to `hdbscan_model` instead of `cruncht_hdbscan_model ` which should contain the BaseCluster. So it might be worthwhile to check whether all the right variables are passed as you intended.

**Comment by annm802** [comment]
> Looking through your code, I noticed you passed `crunch_hdbscan_model` to `hdbscan_model` instead of `cruncht_hdbscan_model ` which should contain the BaseCluster. So it might be worthwhile to check whether all the right variables are passed as you intended.

Oh yeah! It's working now. Thank you so much!
